caas/kubernetes/provider/bootstrap.go

```diff
@@ -0,0 +1,748 @@
+	})
+	return c.broker.updateSecret(secret)
+}
+
+func (c controllerStack) createControllerSecretMongoAdmin() error {
+	return nil
+}
+
+func (c controllerStack) ensureControllerConfigmapBootstrapParams() error {
+	bootstrapParamsFileContent, err := c.pcfg.Bootstrap.StateInitializationParams.Marshal()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	logger.Debugf("bootstrapParams file content: \n%s", string(bootstrapParamsFileContent))
+
+	cm, err := c.getControllerConfigMap()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	cm.Data[c.fileNameBootstrapParams] = string(bootstrapParamsFileContent)
+
+	logger.Debugf("creating bootstrap-params configmap: \n%+v", cm)
+	c.cleanUps = append(c.cleanUps, func() {
+		logger.Debugf("deleting %q bootstrap-params", cm.Name)
+		c.broker.deleteConfigMap(cm.Name)
+	})
+	return c.broker.ensureConfigMap(cm)
+}
+
+func (c controllerStack) ensureControllerConfigmapAgentConf() error {
+	agentConfigFileContent, err := c.agentConfig.Render()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	logger.Debugf("agentConfig file content: \n%s", string(agentConfigFileContent))
+
+	cm, err := c.getControllerConfigMap()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	cm.Data[c.fileNameAgentConf] = string(agentConfigFileContent)
+
+	logger.Debugf("ensuring agent.conf configmap: \n%+v", cm)
+	c.cleanUps = append(c.cleanUps, func() {
+		logger.Debugf("deleting %q template-agent.conf", cm.Name)
+		c.broker.deleteConfigMap(cm.Name)
+	})
+	return c.broker.ensureConfigMap(cm)
+}
+
+func (c controllerStack) createControllerStatefulset() error {
+	numberOfPods := int32(1) // TODO: HA mode!
+	spec := &apps.StatefulSet{
+		ObjectMeta: v1.ObjectMeta{
+			Name:      c.resourceNameStatefulSet,
+			Labels:    c.stackLabels,
+			Namespace: c.namespace,
+		},
+		Spec: apps.StatefulSetSpec{
+			ServiceName: c.resourceNameService,
+			Replicas:    &numberOfPods,
+			Selector: &v1.LabelSelector{
+				MatchLabels: c.stackLabels,
+			},
+			Template: core.PodTemplateSpec{
+				ObjectMeta: v1.ObjectMeta{
+					Labels:    c.stackLabels,
+					Namespace: c.namespace,
+				},
+				Spec: core.PodSpec{
+					RestartPolicy: core.RestartPolicyAlways,
+				},
+			},
+		},
+	}
+
+	if err := c.buildStorageSpecForController(spec); err != nil {
+		return errors.Trace(err)
+	}
+
+	if err := c.buildContainerSpecForController(spec); err != nil {
+		return errors.Trace(err)
+	}
+
+	logger.Debugf("creating controller statefulset: \n%+v", spec)
+	c.cleanUps = append(c.cleanUps, func() {
+		logger.Debugf("deleting %q statefulset", spec.Name)
+		c.broker.deleteStatefulSet(spec.Name)
+	})
+	return errors.Trace(c.broker.createStatefulSet(spec))
+}
+
+func (c controllerStack) buildStorageSpecForController(statefulset *apps.StatefulSet) error {
+	storageclass, err := c.broker.getDefaultStorageClass()
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return errors.Annotatef(err, "no default storageclass found, please create one then retry")
```

Comment:
Add a TODO to fix this.
We need to fail early. We can use the work I am doing to make this more user friendly.

Reply:
added TODO